### ISSUE TITLE
order by desc with backward index scan slow/freezing/crashing

### ISSUE THREAD

**Issue by loheine** [yugabyte/yugabyte-db#4469]
Hi,

I'm running 2.1.6 Yugabyte test cluster on a 8 core ubuntu machine.

Please check the appended YSQL code.

Order by asc is fast 5ms.

Order by desc (last query) is slow and timeouts often after 60secs.

After a couple of retries Yugabytes crashes and won't start up again.

```
drop table events_backwardscan;

create table events_backwardscan (

  log       text not null,
  src       text not null,
  inserted  timestamp(3) without time zone not null,
  created   timestamp(3) without time zone not null,
  data      jsonb not null,
  
  primary key (log, src, created)
);

create index on events_backwardscan (inserted asc);

-- insert 777601 rows (1-2min duration)

insert into events_backwardscan

select
  'log',
  'src',
  t,
  t,
  '{}'
  
from generate_series(
  timestamp '2020-01-01', 
  timestamp '2020-01-10',
  interval  '1 second'
) 

as t(day);

explain analyse
select * 
from events_backwardscan 
order by created asc
limit 100;

explain analyse
select * 
from events_backwardscan 
order by inserted desc
limit 100;
 ```

```
W0514 07:12:26.600299 28732 cluster_balance.cc:227] Skipping load balancing 0000400000003000800000000000405d: Leader not ready to serve requests. (../../src/yb/master/cluster_balance_util.h:286): Master leader has not yet received heartbeat from ts 4170a57782b245b999a2271f87419bde, either master just became leader or a network partition.
W0514 07:12:26.600335 28732 cluster_balance.cc:227] Skipping load balancing 0000400000003000800000000000405e: Leader not ready to serve requests. (../../src/yb/master/cluster_balance_util.h:286): Master leader has not yet received heartbeat from ts 4170a57782b245b999a2271f87419bde, either master just became leader or a network partition.
W0514 07:12:26.600373 28732 cluster_balance.cc:227] Skipping load balancing 0000400000003000800000000000405f: Leader not ready to serve requests. (../../src/yb/master/cluster_balance_util.h:286): Master leader has not yet received heartbeat from ts 4170a57782b245b999a2271f87419bde, either master just became leader or a network partition.
W0514 07:12:26.600440 28732 cluster_balance.cc:227] Skipping load balancing 0000400000003000800000000000406c: Leader not ready to serve requests. (../../src/yb/master/cluster_balance_util.h:286): Master leader has not yet received heartbeat from ts 4170a57782b245b999a2271f87419bde, either master just became leader or a network partition.
W0514 07:12:26.600476 28732 cluster_balance.cc:227] Skipping load balancing 00004000000030008000000000004071: Leader not ready to serve requests. (../../src/yb/master/cluster_balance_util.h:286): Master leader has not yet received heartbeat from ts 4170a57782b245b999a2271f87419bde, either master just became leader or a network partition.
W0514 07:12:26.600569 28732 cluster_balance.cc:227] Skipping load balancing 0000400000003000800000000000408a: Leader not ready to serve requests. (../../src/yb/master/cluster_balance_util.h:286): Master leader has not yet received heartbeat from ts 4170a57782b245b999a2271f87419bde, either master just became leader or a network partition.
W0514 07:12:26.600606 28732 cluster_balance.cc:227] Skipping load balancing 0000400000003000800000000000408f: Leader not ready to serve requests. (../../src/yb/master/cluster_balance_util.h:286): Master leader has not yet received heartbeat from ts 4170a57782b245b999a2271f87419bde, either master just became leader or a network partition.
W0514 07:12:26.600646 28732 cluster_balance.cc:227] Skipping load balancing 157f6a5ea3be49dbbff2ad062ff1d411: Leader not ready to serve requests. (../../src/yb/master/cluster_balance_util.h:286): Master leader has not yet received heartbeat from ts 4170a57782b245b999a2271f87419bde, either master just became leader or a network partition.
W
```

**Comment by loheine** [comment]
I had to restart the cluster.
After restart the first explain slow and not using the index anymore...
![image](https://user-images.githubusercontent.com/61897905/81906389-73f94280-95c6-11ea-982f-68a7c5ae4a1f.png)

**Comment by loheine** [comment]
I had to restart the cluster.

And I rerun all queries.

I've attached the explain results.

![image](https://user-images.githubusercontent.com/61897905/81906808-1dd8cf00-95c7-11ea-9a2b-f520c225a2bf.png)
![image](https://user-images.githubusercontent.com/61897905/81906912-4bbe1380-95c7-11ea-844a-e3e79b70974c.png)

**Comment by ddorian** [comment]
Can replicate the findings above.
After restarting the yb-tserver the query works. (it worked for the user too)

```
yugabyte=# explain analyse
select * 
from events_backwardscan 
order by inserted desc
limit 100;
                                                                                  QUERY PLAN                       
                                                           
-------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------
 Limit  (cost=0.00..10.40 rows=100 width=112) (actual time=5.744..8.712 rows=100 loops=1)
   ->  Index Scan Backward using events_backwardscan_inserted_idx on events_backwardscan  (cost=0.00..104.01 rows=1
000 width=112) (actual time=5.742..8.698 rows=100 loops=1)
 Planning Time: 0.110 ms
 Execution Time: 8.776 ms
(4 rows)
```
It's probably connected to the heavy transaction with scanning intents.

**Comment by m-iancu** [comment]
@bmatican Just fyi, here's a perf stack from testing locally (before restart).  
```
370851	370851.0ms   90.7%	0.0	 	                yb::docdb::IntentAwareIterator::PrevDocKey(yb::Slice const&)
370851	370851.0ms   90.7%	0.0	 	                 yb::docdb::IntentAwareIterator::PreparePrev(yb::Slice const&)
356753	356753.0ms   87.2%	4955.0	 	                  rocksdb::DBIter::SeekToLast()
351798	351798.0ms   86.0%	18079.0	 	                   rocksdb::DBIter::PrevInternal()
332092	332092.0ms   81.2%	10777.0	 	                    rocksdb::DBIter::FindValueForCurrentKey()
318454	318454.0ms   77.9%	59122.0	 	                     rocksdb::MemTableIterator::Prev()
259332	259332.0ms   63.4%	55722.0	 	                      rocksdb::(anonymous namespace)::SkipListRep<rocksdb::SingleWriterInlineSkipList<rocksdb::MemTableRep::KeyComparator const&> >::Iterator::Prev()
203610	203610.0ms   49.8%	107758.0	 	                       rocksdb::MemTable::KeyComparator::operator()(char const*, char const*) const
95852	95852.0ms   23.4%	95852.0	 	                        rocksdb::InternalKeyComparator::Compare(yb::Slice const&, yb::Slice const&) const
2861	2861.0ms    0.6%	2539.0	 	                     rocksdb::DBIter::ParseKey(rocksdb::ParsedInternalKey*)
```
Waiting for a few minutes after the data load did not help. But restarting the tserver as @ddorian reported fixed it.

**Comment by mbautin** [comment]
In my test based on the above schema/data/statement, I see that PrevInternal in db_iter.cc loops 170000 times on every iteration. The intents RocksDB is empty, but probably contains a lot of delete markers from removed intents, and PrevInternal (called as part of SeekToLast) loops through all of them on every iteration. I am working on a fix.

**Comment by mbautin** [comment]
Verified that this issue is fixed by ae23f9955caf591185c2e9c6e81ddf7d8299d8e6. Adding a new unit test for that.

**Comment by mbautin** [comment]
The issue described here should be fixed by ae23f9955caf591185c2e9c6e81ddf7d8299d8e6, but I've created #4672 to follow up on the theoretically possible but not easily reproducible case of running into a similar case if there are pending transactions.
